File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/library/items/media/files/nowplaying/NowPlayingFloatingActionButton.java
package com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.nowplaying;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.view.ViewGroup;
import android.widget.RelativeLayout;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.lasthopesoftware.bluewater.R;
import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.nowplaying.activity.NowPlayingActivity;
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents;
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils;
import com.namehillsoftware.handoff.promises.response.VoidResponse;

public class NowPlayingFloatingActionButton extends FloatingActionButton {
    public static NowPlayingFloatingActionButton addNowPlayingFloatingActionButton(RelativeLayout container) {
        final NowPlayingFloatingActionButton nowPlayingFloatingActionButton = new NowPlayingFloatingActionButton(container.getContext());

        final RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        layoutParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
        layoutParams.addRule(RelativeLayout.ALIGN_PARENT_END);
        final int margin = ViewUtils.dpToPx(container.getContext(), 16);
        layoutParams.setMargins(margin, margin, margin, margin);

        nowPlayingFloatingActionButton.setLayoutParams(layoutParams);
        container.addView(nowPlayingFloatingActionButton);

        return nowPlayingFloatingActionButton;
    }

    private boolean isNowPlayingFileSet;

    private NowPlayingFloatingActionButton(Context context) {
        super(context);

        setImageDrawable(ViewUtils.getDrawable(context, R.drawable.av_play_dark));

        initializeNowPlayingFloatingActionButton();
    }


    @SuppressWarnings("ResourceType")
    private void initializeNowPlayingFloatingActionButton() {
        setOnClickListener(v -> NowPlayingActivity.startNowPlayingActivity(v.getContext()));

        setVisibility(ViewUtils.getVisibility(false));
        // The user can change the library, so let's check if the state of visibility on the
        // now playing menu item should change
        NowPlayingFileProvider
            .fromActiveLibrary(getContext())
            .getNowPlayingFile()
            .then(new VoidResponse<>(result -> {
                isNowPlayingFileSet = result != null;
                setVisibility(ViewUtils.getVisibility(isNowPlayingFileSet));

                if (isNowPlayingFileSet) return;

                final LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.getInstance(getContext());

                localBroadcastManager.registerReceiver(new BroadcastReceiver() {
                    @Override
                    public synchronized void onReceive(Context context, Intent intent) {
                        isNowPlayingFileSet = true;
                        setVisibility(ViewUtils.getVisibility(true));
                        localBroadcastManager.unregisterReceiver(this);
                    }
                }, new IntentFilter(PlaylistEvents.onPlaylistStart));
            }));
    }

    @Override
    public void show() {
        if (isNowPlayingFileSet) super.show();
    }

    @Override
    public void show(OnVisibilityChangedListener listener) {
        if (isNowPlayingFileSet) super.show(listener);
    }
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/view/nowplaying/NowPlayingFloatingActionButton.kt
package com.lasthopesoftware.bluewater.client.playback.view.nowplaying

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.NowPlayingFileProvider.Companion.fromActiveLibrary
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.activity.NowPlayingActivity.Companion.startNowPlayingActivity
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils

class NowPlayingFloatingActionButton private constructor(context: Context) : FloatingActionButton(context) {
	private var isNowPlayingFileSet = false

	private fun initializeNowPlayingFloatingActionButton() {
		setOnClickListener { v -> startNowPlayingActivity(v.context) }

		visibility = ViewUtils.getVisibility(false)
		// The user can change the library, so let's check if the state of visibility on the
		// now playing menu item should change

		fromActiveLibrary(context)
			?.nowPlayingFile
			?.then { result ->
				isNowPlayingFileSet = result != null
				visibility = ViewUtils.getVisibility(isNowPlayingFileSet)
				if (isNowPlayingFileSet) return@then

				val localBroadcastManager = LocalBroadcastManager.getInstance(context)
				localBroadcastManager.registerReceiver(object : BroadcastReceiver() {
					@Synchronized
					override fun onReceive(context: Context, intent: Intent) {
						isNowPlayingFileSet = true
						visibility = ViewUtils.getVisibility(true)
						localBroadcastManager.unregisterReceiver(this)
					}
				}, IntentFilter(PlaylistEvents.onPlaylistStart))
			}
	}

	override fun show() {
		if (isNowPlayingFileSet) super.show()
	}

	override fun show(listener: OnVisibilityChangedListener?) {
		if (isNowPlayingFileSet) super.show(listener)
	}

	companion object {
		@JvmStatic
		fun addNowPlayingFloatingActionButton(container: RelativeLayout): NowPlayingFloatingActionButton {
			val nowPlayingFloatingActionButton = NowPlayingFloatingActionButton(container.context)
			val layoutParams = RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
			layoutParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM)
			layoutParams.addRule(RelativeLayout.ALIGN_PARENT_END)
			val margin = ViewUtils.dpToPx(container.context, 16)
			layoutParams.setMargins(margin, margin, margin, margin)
			nowPlayingFloatingActionButton.layoutParams = layoutParams
			container.addView(nowPlayingFloatingActionButton)
			return nowPlayingFloatingActionButton
		}
	}

	init {
		setImageDrawable(ViewUtils.getDrawable(context, R.drawable.av_play_dark))
		initializeNowPlayingFloatingActionButton()
	}
}
